src/hosts.rs

```diff
@@ -0,0 +1,323 @@
+        self.status
+    }
+
+    /// Checks if the host is online.
+    ///
+    /// # Returns
+    ///
+    /// true if the host is online, false otherwise.
+    pub fn is_online(&self) -> bool {
+        self.status == HostStatus::Online
+    }
+
+    /// Checks if the Readyset host is ready to serve traffic.
+    /// This is done by querying the SHOW READYSET STATUS command.
+    ///
+    /// # Returns
+    ///
+    /// true if the host is ready, false otherwise.
+    pub fn check_readyset_is_ready(&mut self) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            Some(conn) => {
+                let rows: Vec<(String, String)> =
+                    conn.query("SHOW READYSET STATUS").unwrap_or(vec![]);
+                for (field, value) in rows {
+                    if field == "Snapshot Status" {
+                        return Ok(value == "Completed");
+                    }
+                }
+                Ok(false)
+            }
+            None => Err(mysql::Error::IoError(std::io::Error::new(
+                std::io::ErrorKind::Other,
+                "Connection to Readyset host is not established",
+            ))),
+        }
+    }
+
+    /// Checks if the host supports the given query.
+    /// This is done by querying the EXPLAIN CREATE CACHE FROM command.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest_text` - The digest text of the query.
+    /// * `schema` - The schema of the query.
+    ///
+    /// # Returns
+    ///
+    /// true if the host supports the query, false otherwise.
+    pub fn check_query_support(
+        &mut self,
+        digest_text: &String,
+        schema: &String,
+    ) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            Some(conn) => {
+                conn.query_drop(format!("USE {}", schema))
+                    .expect("Failed to use schema");
+                let row: Option<(String, String, String)> =
+                    conn.query_first(format!("EXPLAIN CREATE CACHE FROM {}", digest_text))?;
+                match row {
+                    Some((_, _, value)) => Ok(value == "yes" || value == "cached"),
+                    None => Ok(false),
+                }
+            }
+            None => Ok(false),
+        }
+    }
+
+    /// Caches the given query on the host.
+    /// This is done by executing the CREATE CACHE FROM command.
+    ///
+    /// # Arguments
+    ///
+    /// * `digest_text` - The digest text of the query.
+    ///
+    /// # Returns
+    ///
+    /// true if the query was cached successfully, false otherwise.
+    pub fn cache_query(
+        &mut self,
+        digest_text: &String,
+        digest: &String,
+    ) -> Result<bool, mysql::Error> {
+        match &mut self.conn {
+            None => return Ok(false),
+            Some(conn) => {
+                conn.query_drop(format!("CREATE CACHE d_{} FROM {}", digest, digest_text))
+                    .expect("Failed to create readyset cache");
```

Comment:
I think this could fail for a lot of unexpected reasons and we should probably return an error instead of panic.

Reply:
Fixed. Btw, we are panicking  in the caller if cache_query returns an error. Basically we are iterating through all online readyset hosts and if one readyset fails to cache a query, this is one unrecoverable error. Given the nature of the scheduler (crontab that runs every X seconds) I don't see any difference in panicking right away versus handling an error.